src/renderers/js-experimental/fragments/programInstructions.ts

```diff
@@ -103,3 +105,40 @@ function getProgramInstructionsIdentifierFunctionFragment(
       `}`
   );
 }
+
+function getProgramUnionParsedInstructionTypeFragment(
+  scope: Pick<GlobalFragmentScope, 'nameApi'> & {
+    programNode: ProgramNode;
+    allInstructions: InstructionNode[];
+  }
+): Fragment {
+  const { programNode, allInstructions, nameApi } = scope;
+
+  const programInstructionsType = nameApi.programInstructionsTypeUnion(
+    programNode.name
+  );
+
+  const programInstructionsEnum = nameApi.programInstructionsEnum(
+    programNode.name
+  );
+
+  const typeVariants = allInstructions.map((instruction): Fragment => {
+    const instructionEnumVariant = nameApi.programInstructionsEnumVariant(
+      instruction.name
+    );
+
+    const parsedInstructionType = nameApi.instructionParsedType(
+      instruction.name
+    );
+
+    return fragment(
+      `| {instructionType: ${programInstructionsEnum}.${instructionEnumVariant}} & ${parsedInstructionType}`,
+      new ImportMap().add('../instructions', parsedInstructionType)
+    );
```

Comment:
Two small things here:
- You can just chain methods on `Fragment` objects such as `addImports`.
- It's best to introduce a new `"generatedInstructions"` alias on the `ImportMap` so we always have a single place to update these folders (if we ever need to).

```suggestion
    return fragment(`| { instructionType: ${programInstructionsEnum}.${instructionEnumVariant} } & ${parsedInstructionType}`)
      .addImports('generatedInstructions', parsedInstructionType);
```